lib/internal/modules/esm/get_format.js

```diff
@@ -54,30 +56,48 @@ const protocolHandlers = ObjectAssign(ObjectCreate(null), {
     return format;
   },
   'file:'(parsed, url) {
-    const ext = extname(parsed.pathname);
-    let format;
-
-    if (ext === '.js') {
-      format = getPackageType(parsed.href) === 'module' ? 'module' : 'commonjs';
-    } else {
-      format = extensionFormatMap[ext];
-    }
-    if (!format) {
-      if (experimentalSpecifierResolution === 'node') {
-        process.emitWarning(
-          'The Node.js specifier resolution in ESM is experimental.',
-          'ExperimentalWarning');
-        format = legacyExtensionFormatMap[ext];
-      } else {
-        throw new ERR_UNKNOWN_FILE_EXTENSION(ext, fileURLToPath(url));
-      }
-    }
-
-    return format || null;
+    return getModuleFileFormat(parsed, true, url) || null;
   },
   'node:'() { return 'builtin'; },
 });
 
+function throwIfNotExperimentalSpecifierResolution(ext, url) {
+  if (experimentalSpecifierResolution !== 'node') {
+    throw new ERR_UNKNOWN_FILE_EXTENSION(ext, fileURLToPath(url));
+  }
+  return getLegacyExtensionFormat(ext);
+}
+
+function getLegacyExtensionFormat(ext) {
+  if (
+    experimentalSpecifierResolution === 'node' &&
+    !experimentalSpecifierResolutionWarned
+  ) {
+    process.emitWarning(
+      'The Node.js specifier resolution in ESM is experimental.',
+      'ExperimentalWarning');
+    experimentalSpecifierResolutionWarned = true;
+  }
+  return legacyExtensionFormatMap[ext];
+}
+
+function getModuleFileFormat(url, throwIfNotResolved, urlForThrow) {
+  let format;
+
+  const ext = extname(url.pathname);
+  if (ext === '.js') {
+    format = getPackageType(url) === 'module' ? 'module' : 'commonjs';
```

Comment:
if the package type points to something that's not "module" or "commonjs", should this throw, rather than silently/wrongly being commonjs?

Reply:
that is a good point but it was like that before this PR and I would not change that behaviour here. As far as I understood it seems to be an older convention that for `.js` files everything that is not `module` has to be `commonjs`.

Reply:
For backward compatibility, we _have to_ ignore it in case a `package.json` was created before Node.js introduced this field. This decision can be re-evaluated at some point with a deprecation cycle, but it's out of scope for this PR.

Reply:
I think if we did throw, that throw should be deeper down, perhaps in `getPackageType()`, and would have far broader impact/scope than this PR.

@aduh95 was the thinking at the time that people could have been using the `type` field for something else? If so, could we not add the throw as a breaking change in a major release and then those package authors can add an `"engines"` check if they don't want to update their very, very outdated use of `"type"`?

Reply:
> was the thinking at the time that people could have been using the `type` field for something else?

I suppose, I wasn't part of these discussions. 

> If so, could we not add the throw as a breaking change in a major release and then those package authors can add an `"engines"` check if they don't want to update their very, very outdated use of `"type"`?

My opinion on this is we should wait for loaders to be stable, and then add a runtime warning for all packages that do not contain a valid `"type"` field, then after a bit switch the default to ESM. Not sure this is actually feasible (or desirable) without breaking all legacy packages though.

Reply:
That makes perfect sense to me as reasoning and as a potential path.

I wasn't thinking of the "type" field tho - just the return type of `getPackageType`.